File: M3Tools/Types/Orders/Order.vb
Option Strict On
Namespace Types
    ' TODO: Consolidate CurrentOrder and CompletedOrder
    Public Class Order
        Inherits DBEntry

		<Text.Json.Serialization.JsonPropertyName("customer")>
		Public Property Customer As Person

		<Text.Json.Serialization.JsonPropertyName("item")>
		Public Property Item As Item

		<Text.Json.Serialization.JsonPropertyName("quantity")>
		Public Property Quantity As Integer

		<Text.Json.Serialization.JsonPropertyName("total")>
		Public ReadOnly Property OrderTotal As Double
			Get
				Return Item.Price * Quantity
			End Get
		End Property

		<Text.Json.Serialization.JsonPropertyName("ordered")>
		Public Property OrderDate As Date

		<Text.Json.Serialization.JsonPropertyName("completed")>
		Public Property CompletedDate As Date

		Public Sub New()
			Me.New(-1, -1, -1, 0)
		End Sub

		''' <summary>
		''' New CurrentOrder Object
		''' </summary>
		''' <param name="orderID"></param>
		''' <param name="customerID"></param>
		''' <param name="itemID"></param>
		''' <param name="quantity"></param>
		''' <param name="orderDate"></param>
		Public Sub New(orderID As Integer, customerID As Integer, itemID As Integer, quantity As Integer, Optional orderDate As Date = Nothing, Optional completedDate As Date = Nothing)
			MyBase.New(orderID)
			Me.Quantity = quantity
			Me.OrderDate = If(orderDate.Year < 2000, Date.Now, orderDate)
			Me.CompletedDate = completedDate
			GetCustomer(customerID)
			GetItem(itemID)
		End Sub

		''' <summary>
		''' Retrieve a user from the database using their CustomerID
		''' </summary>
		''' <param name="customerID">CustomerID of the desired customer</param>
		Private Sub GetCustomer(customerID As Integer)
			If Not Utils.ValidID(customerID) Then
				Exit Sub
			End If

			Using c As New Database.CustomerDatabase
				Customer = c.GetCustomer(customerID)
			End Using
		End Sub

		''' <summary>
		''' Retrieve an item from the database using the provided itemID
		''' </summary>
		''' <param name="itemID">ItemID of the desired item</param>
		Private Sub GetItem(itemID As Integer)
			If Not Utils.ValidID(itemID) Then
				Exit Sub
			End If

			Using i As New Database.InventoryDatabase
				Item = i.GetItem(itemID)
			End Using
		End Sub

		'Public Overrides Sub UpdateID(newID As Integer)
		'	If newID = Id Then
		'		Return
		'	End If

		'	Using conn As New Database.OrdersDatabase
		'		Dim newOrder = conn.GetOrderById(newID)

		'		' TODO: Finish implementing updates
		'	End Using
		'End Sub

		Public Function Clone() As Order
			Return New Order(Id, Customer.Id, Item.Id, Quantity, OrderDate, CompletedDate)
		End Function
	End Class
End Namespace

File: M3Tools/Types/Orders/Order.cs
using System;

namespace SPPBC.M3Tools.Types
{
	// TODO: Consolidate CurrentOrder and CompletedOrder
	/// <summary>
	/// 
	/// </summary>
	public class Order : DbEntry
	{

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
		[Newtonsoft.Json.JsonProperty("orderID")]
		public override int Id
		{
			get => base.Id;
			set => base.Id = value;
		}

		/// <summary>
		/// The customer the order was placed for
		/// </summary>
		[Newtonsoft.Json.JsonProperty("customer")]
		public Customer Customer { get; set; }

		/// <summary>
		/// The item the order was placed for
		/// </summary>
		[Newtonsoft.Json.JsonProperty("item")]
		public Product Item { get; set; }

		/// <summary>
		/// How many of the item the customer wanted
		/// </summary>
		[Newtonsoft.Json.JsonProperty("quantity")]
		public int Quantity { get; set; }

		/// <summary>
		/// The subtotal for the order
		/// </summary>
		[Newtonsoft.Json.JsonProperty("total")]
		public double Total => (double)(Item.Price * Quantity);

		/// <summary>
		/// The date the order was placed
		/// </summary>
		[Newtonsoft.Json.JsonProperty("ordered")]
		public DateTime OrderDate { get; set; }

		/// <summary>
		/// The date the order was fulfilled
		/// </summary>
		[Newtonsoft.Json.JsonProperty("completed")]
		public DateTime CompletedDate { get; set; }

		/// <summary>
		/// An empty instance of an Order object
		/// </summary>
		public static new Order None { get; } = new();

		/// <summary>
		/// 
		/// </summary>
		public Order() : this(-1)
		{
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="orderID"></param>
		public Order(int orderID) : this(orderID, -1, -1, 0, default, default)
		{

		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="orderID"></param>
		/// <param name="customerID"></param>
		/// <param name="itemID"></param>
		/// <param name="quantity"></param>
		/// <param name="orderDate"></param>
		/// <param name="completedDate"></param>
		public Order(int orderID, int customerID, int itemID, int quantity, DateTime orderDate = default, DateTime completedDate = default) : this(orderID, new Customer(customerID), new Product(itemID), quantity, orderDate, completedDate)
		{

		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="orderID"></param>
		/// <param name="customer"></param>
		/// <param name="item"></param>
		/// <param name="quantity"></param>
		/// <param name="orderDate"></param>
		/// <param name="completedDate"></param>
		public Order(int orderID, Customer customer, Product item, int quantity, DateTime orderDate, DateTime completedDate) : base(orderID)
		{
			Customer = customer;
			Item = item;
			Quantity = quantity;
			OrderDate = orderDate.Year > 2000 ? orderDate : DateTime.Now;
			CompletedDate = completedDate.Year > 2000 ? completedDate : default;
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="left"></param>
		/// <param name="right"></param>
		/// <returns></returns>
		public static bool operator ==(Order left, Order right) => !(left is null ^ right is null) && left.Id == right.Id && left.Customer == right.Customer && left.Item == right.Item && left.Quantity == right.Quantity && left.OrderDate == right.OrderDate && left.CompletedDate == right.CompletedDate;

		/// <summary>
		/// 
		/// </summary>
		/// <param name="left"></param>
		/// <param name="right"></param>
		/// <returns></returns>
		public static bool operator !=(Order left, Order right) => !(left == right);

		/// <inheritdoc/>
		public override bool Equals(object obj) => this == (obj as Order);

		/// <inheritdoc/>
		public override int GetHashCode() => base.GetHashCode();

		/// <summary>
		/// Clones a copy of the current Order object
		/// </summary>
		/// <returns></returns>
		public new Order Clone() => new(Id, Customer.Id, Item.Id, Quantity, OrderDate, CompletedDate);
	}
}
